### PR TITLE
feat: add Ed25519 signature verification package

### PR DESCRIPTION
**Pull request by junho226**
## Summary
- Add `pkg/signing/keymanager.go`: fetches and caches Ed25519 public key from AI server (`GET /api/commands/public-key/`)
- Add `pkg/signing/verifier.go`: builds canonical payload and verifies Ed25519 signatures
- Add `KeyID` field to `Command` struct for kid-based key cache lookup
- Canonical payload uses struct with alphabetically ordered JSON tags to guarantee byte-identical output with Python's `json.dumps(sort_keys=True, separators=(',', ':'))`

### Key management
- **kid-based lookup**: `GetPublicKeyForKID(kid)` returns cached key if kid matches and not expired, otherwise fetches from AI server
- **expires_at support**: AI server can include `expires_at` in public key response for proactive refresh before expiry, with TTL-based fallback
- **Concurrent refresh guard**: `refreshMu` serializes refresh calls to prevent fetch bursts under load
- **Expired key rejection**: expired keys are never returned on refresh failure

### Canonical payload
- `SetEscapeHTML(false)` to match Python's `json.dumps`
- U+2028/U+2029 unescaping with backslash counting to handle edge cases (Go escapes these, Python does not)
- Nil command checks on public API boundaries

### Defense in depth
- Response body limited to 16KB via `io.LimitReader` with explicit oversized detection
- Empty `key_id` from AI server rejected before cache update

## Test coverage (25 tests)
- Key fetch, caching, TTL expiry, expires_at expiry
- kid-based cache hit, kid mismatch, key rotation
- Expired key + refresh failure → error
- Canonical payload: basic, empty fields, HTML chars, U+2028/U+2029, literal backslash-u
- Signature: valid, tampered, wrong server ID, empty, invalid base64, wrong size, nil key, nil command
- Server errors: unavailable, invalid algorithm, invalid key size

## Cross-repo compatibility
- **alpacon-protos** (PR #23): `CommandExecute` proto adds `signature`, `key_id`, `analyzed_at` fields
- **alpacon-server** (PR #1816): stores `key_id` from AI server, includes signing fields in command dispatch
- **alpacon-ai-server**: verify response includes `key_id`, public-key response includes `expires_at`

## Test plan
- [x] All 25 signing package tests pass
- [x] All CI checks pass (golangci-lint, build, 19 distros)
- [ ] Cross-language test: sign payload in Python, verify in Go (to be validated during integration)

Stacked on #263

Refs #262

### CODE REVIEW

File: pkg/signing/verifier.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+		CommandID: cmd.ID,
+		GroupName: cmd.Group,
+		Line:      cmd.Line,
+		ServerID:  serverID,
+		Shell:     cmd.Shell,
+		Timestamp: cmd.AnalyzedAt,
+		Username:  cmd.User,
+	}
+	data, _ := json.Marshal(p)
+	return data
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/verifier.go]: Fixed — switched to `json.Encoder` with `SetEscapeHTML(false)` to match Python behavior. Added test with `<`, `>`, and `&` characters.

File: pkg/signing/verifier.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	}
+
+	if len(sig) != ed25519.SignatureSize {
+		return fmt.Errorf("invalid signature size: got %d, want %d", len(sig), ed25519.SignatureSize)
+	}
+
+	payload := BuildCanonicalPayload(cmd, serverID)
+
+	if !ed25519.Verify(publicKey, payload, sig) {
+		return errors.New("signature verification failed")
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/verifier.go:121]: Fixed — added public key size validation before calling `ed25519.Verify`. Added test for nil public key.

File: pkg/signing/keymanager.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	}
+}
+
+// GetPublicKey returns the cached public key, refreshing if stale.
+func (m *KeyManager) GetPublicKey() (ed25519.PublicKey, error) {
+	m.mu.RLock()
+	if m.publicKey != nil && time.Since(m.lastFetch) < time.Duration(m.refreshSecs)*time.Second {
+		key := m.publicKey
+		m.mu.RUnlock()
+		return key, nil
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/keymanager.go:68]: Fixed — `GetPublicKey()` now returns a defensive copy via `copyKey()` helper.

File: pkg/signing/keymanager.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	return m.publicKey, nil
+}
+
+// Refresh fetches the latest public key from the AI server.
+func (m *KeyManager) Refresh() error {
+	url := m.aiBaseURL + "/api/commands/public-key/"
+
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/keymanager.go:165]: Fixed — `NewKeyManager` now trims trailing slashes from `aiBaseURL` in the constructor.

File: pkg/signing/keymanager.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return fmt.Errorf("failed to create request: %w", err)
+	}
+
+	client := &http.Client{
+		Transport: m.client.Transport,
+		Timeout:   m.fetchTimeout,
+	}
+
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/keymanager.go]: Fixed — `NewKeyManager` defaults nil client to `http.DefaultClient`. The derived client now preserves `Transport`, `CheckRedirect`, and `Jar` from the base client. Note: Go stdlib `http.Client` does not have a `Clone()` method.

File: pkg/signing/keymanager_test.go [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+		json.NewEncoder(w).Encode(resp)
+	}))
+	defer server.Close()
+
+	km := NewKeyManager(server.URL, 3600, server.Client())
+
+	// First call fetches
+	km.GetPublicKey()
+	// Second call should use cache
+	km.GetPublicKey()
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/keymanager_test.go]: Fixed — both `GetPublicKey` calls now assert nil errors.

File: pkg/signing/keymanager_test.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+		w.Header().Set("Content-Type", "application/json")
+		json.NewEncoder(w).Encode(resp)
+	}))
+	defer server.Close()
+
+	km := NewKeyManager(server.URL, 1, server.Client()) // 1 second TTL
+
+	km.GetPublicKey()
+	time.Sleep(1100 * time.Millisecond)
+	km.GetPublicKey()
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/keymanager_test.go]: Fixed — both `GetPublicKey` calls now assert nil errors.

File: pkg/signing/verifier.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+		Shell:     cmd.Shell,
+		Timestamp: cmd.AnalyzedAt,
+		Username:  cmd.User,
+	}
+	var buf bytes.Buffer
+	enc := json.NewEncoder(&buf)
+	enc.SetEscapeHTML(false)
+	_ = enc.Encode(p)
+	// Encoder.Encode appends a newline; trim it for canonical form
+	return bytes.TrimRight(buf.Bytes(), "\n")
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/verifier.go]: Fixed — added explicit documentation that both sides use raw UTF-8 (Python's `ensure_ascii=False`, Go's `SetEscapeHTML(false)`). The AI server signing contract specifies `ensure_ascii=False`, so both sides emit identical UTF-8 bytes.

File: pkg/signing/verifier.go [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+// VerifyCommand verifies the Ed25519 signature on a command.
+func VerifyCommand(cmd *protocol.Command, serverID string, publicKey ed25519.PublicKey) error {
+	if len(publicKey) != ed25519.PublicKeySize {
+		return fmt.Errorf("invalid public key size: got %d, want %d", len(publicKey), ed25519.PublicKeySize)
+	}
+
+	if cmd.Signature == "" {
+		return errors.New("empty signature")
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/verifier.go:106]: Fixed — added explicit nil checks to both `VerifyCommand` and `BuildCanonicalPayload`. Tests added.

File: pkg/signing/keymanager.go [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	var keyResp publicKeyResponse
+	if err := json.Unmarshal(body, &keyResp); err != nil {
+		return fmt.Errorf("failed to parse public key response: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/keymanager.go:195]: Fixed — now uses `io.LimitReader(resp.Body, 16KB)` to cap response size.

File: pkg/signing/keymanager.go [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+	m.mu.RUnlock()
+
+	if err := m.Refresh(); err != nil {
+		m.mu.RLock()
+		defer m.mu.RUnlock()
+		if m.publicKey != nil {
+			return copyKey(m.publicKey), nil
+		}
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/keymanager.go:79]: Fixed — added `refreshMu` to serialize refresh calls. `Refresh()` double-checks cache validity after acquiring the lock. `GetPublicKeyForKID()` uses `fetchKey()` which skips the double-check since the key may not be expired but the kid doesn't match.

File: pkg/signing/keymanager_test.go [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+		json.NewEncoder(w).Encode(resp)
+	}))
+	defer server.Close()
+
+	km := NewKeyManager(server.URL, 1, server.Client()) // 1 second TTL
+
+	if _, err := km.GetPublicKey(); err != nil {
+		t.Fatalf("first GetPublicKey failed: %v", err)
+	}
+	time.Sleep(1100 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/keymanager_test.go]: Fixed — replaced `time.Sleep` with direct `km.lastFetch` manipulation to simulate cache expiry.

File: pkg/signing/keymanager_test.go [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+	km := NewKeyManager(server.URL, 1, server.Client()) // 1 second TTL
+
+	// First fetch succeeds
+	key1, err := km.GetPublicKey()
+	if err != nil {
+		t.Fatalf("first fetch should succeed: %v", err)
+	}
+
+	// Wait for cache expiry, next fetch fails but returns stale key
+	time.Sleep(1100 * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/keymanager_test.go]: Fixed — replaced `time.Sleep` with direct `km.lastFetch` manipulation to simulate cache expiry.

File: pkg/signing/verifier.go [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+	Shell     string `json:"shell"`
+	Timestamp string `json:"timestamp"`
+	Username  string `json:"username"`
+}
+
+// BuildCanonicalPayload constructs the signing payload that must match
+// what the AI server signed. The output is deterministic canonical JSON.
+// Uses json.Encoder with SetEscapeHTML(false) to match Python's json.dumps
+// behavior (called with ensure_ascii=False), which does not escape <, >, &
+// or non-ASCII characters. Both sides emit raw UTF-8.
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/verifier.go]: Fixed — Go's encoding/json escapes U+2028/U+2029 even with SetEscapeHTML(false). Added post-processing to unescape these to match Python's `ensure_ascii=False` output. Added test to lock down the contract.

File: pkg/signing/keymanager.go [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+	defer func() { _ = resp.Body.Close() }()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("public key endpoint returned status %d", resp.StatusCode)
+	}
+
+	body, err := io.ReadAll(io.LimitReader(resp.Body, maxResponseSize))
+	if err != nil {
+		return fmt.Errorf("failed to read response body: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/keymanager.go]: Fixed — now reads `maxResponseSize+1` bytes and returns an explicit error if the response exceeds the limit, preventing silent truncation.

File: pkg/signing/keymanager.go [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+func (m *KeyManager) GetPublicKeyForKID(kid string) (ed25519.PublicKey, error) {
+	m.mu.RLock()
+	if m.publicKey != nil && m.keyID == kid {
+		key := copyKey(m.publicKey)
+		m.mu.RUnlock()
+		return key, nil
+	}
+	m.mu.RUnlock()
+
+	// Unknown kid: force fetch new key from AI server
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/keymanager.go]: Fixed — added `!m.isExpired()` check to the cache-hit path in `GetPublicKeyForKID`, so expired keys trigger a refresh even when the kid matches.

File: pkg/signing/verifier.go [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+	enc := json.NewEncoder(&buf)
+	enc.SetEscapeHTML(false)
+	_ = enc.Encode(p)
+	// Encoder.Encode appends a newline; trim it for canonical form
+	result := bytes.TrimRight(buf.Bytes(), "\n")
+	// Unescape U+2028/U+2029 that Go escapes but Python does not
+	result = []byte(strings.NewReplacer(
+		`\u2028`, "\u2028",
+		`\u2029`, "\u2029",
+	).Replace(string(result)))
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/verifier.go]: Fixed — replaced naive `strings.NewReplacer` with `unescapeLineSeparators()` that scans bytes and only replaces `\u2028`/`\u2029` when the backslash is not itself escaped. Literal `\\u2028` in user input is preserved correctly.

File: pkg/signing/verifier_test.go [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+	// Payload must contain raw UTF-8 bytes, not \u2028/\u2029 escapes
+	if bytes.Contains(payload, []byte(`\u2028`)) || bytes.Contains(payload, []byte(`\u2029`)) {
+		t.Errorf("U+2028/U+2029 should not be escaped in canonical payload\ngot: %s", string(payload))
+	}
+	// Verify the raw bytes are present
+	if !bytes.Contains(payload, []byte("\u2028")) || !bytes.Contains(payload, []byte("\u2029")) {
+		t.Error("payload should contain raw U+2028/U+2029 bytes")
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/verifier_test.go:231]: Fixed — added `TestBuildCanonicalPayload_LiteralBackslashU2028` that verifies literal `\u2028`/`\u2029` text in user input is not corrupted into raw bytes.

File: pkg/signing/verifier.go [hunk 19]
```diff
@@ -0,0 +1,114 @@
+package signing
+
+import (
+	"bytes"
+	"crypto/ed25519"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/verifier.go:9]: Fixed — added blank line between stdlib and third-party import groups.

File: pkg/signing/verifier.go [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+// unescapeLineSeparators replaces JSON-escaped \u2028 and \u2029 with raw
+// UTF-8 bytes, but only when the backslash is not itself escaped (\\u2028).
+func unescapeLineSeparators(data []byte) []byte {
+	var buf bytes.Buffer
+	buf.Grow(len(data))
+	for i := 0; i < len(data); i++ {
+		if data[i] == '\\' && i+5 < len(data) && data[i+1] == 'u' &&
+			(string(data[i+2:i+6]) == "2028" || string(data[i+2:i+6]) == "2029") {
+			// Check if the backslash is escaped (preceded by another backslash)
+			if i > 0 && data[i-1] == '\\' {
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/verifier.go]: Fixed — replaced single backslash check with consecutive preceding backslash counting. Odd count means the backslash is escaped, even means it's a real JSON escape. Handles `\\\u2028` correctly.

File: pkg/signing/keymanager.go [hunk 21]
```diff
@@ -0,0 +1,228 @@
...
+	m.mu.RUnlock()
+
+	// Unknown kid or expired key: force fetch new key from AI server
+	if err := m.fetchKey(); err != nil {
+		m.mu.RLock()
+		defer m.mu.RUnlock()
+		if m.publicKey != nil {
+			return copyKey(m.publicKey), nil
+		}
+		return nil, err
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/keymanager.go]: Fixed — `GetPublicKeyForKID` fallback now only returns the cached key if `m.keyID == kid && !m.isExpired()`, otherwise returns a wrapped error.

File: pkg/signing/keymanager.go [hunk 22]
```diff
@@ -0,0 +1,228 @@
...
+
+	if len(keyBytes) != ed25519.PublicKeySize {
+		return fmt.Errorf("invalid public key size: got %d, want %d", len(keyBytes), ed25519.PublicKeySize)
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.publicKey = ed25519.PublicKey(keyBytes)
+	m.keyID = keyResp.KeyID
+	m.lastFetch = time.Now()
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/keymanager.go:218]: Fixed — `fetchKeyLocked` now rejects empty `key_id` from the AI server response with an explicit error before updating the cache.

File: pkg/signing/keymanager.go [hunk 23]
```diff
@@ -0,0 +1,234 @@
...
+		return key, nil
+	}
+	m.mu.RUnlock()
+
+	if err := m.Refresh(); err != nil {
+		m.mu.RLock()
+		defer m.mu.RUnlock()
+		if m.publicKey != nil {
+			return copyKey(m.publicKey), nil
+		}
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/keymanager.go]: Fixed — `GetPublicKey` now only returns cached key on refresh failure if it's not expired. Expired key + refresh failure returns an error.

File: pkg/signing/keymanager.go [hunk 24]
```diff
@@ -0,0 +1,234 @@
...
+		return nil, fmt.Errorf("failed to refresh key for kid %q: %w", kid, err)
+	}
+
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	if m.keyID != kid {
+		return nil, fmt.Errorf("key_id mismatch after refresh: cached %q, requested %q", m.keyID, kid)
+	}
+	return copyKey(m.publicKey), nil
```
[the remark this answers is not in the document]
Comment (junho226) [reply pkg/signing/keymanager.go]: Fixed — post-refresh check now uses `m.keyID == kid && !m.isExpired()` for a consistent result, avoiding race conditions from concurrent refreshes overwriting the cache.

**Review by eunyoung14 — DISMISSED** [review]
LGTM. Clean implementation with good test coverage (25 tests).

kid-based key management, concurrent refresh guard, and expires_at support are well designed. The U+2028/U+2029 unescaping for Go/Python canonical JSON compatibility is correctly handled with backslash counting.

Minor notes (non-blocking):
- `fetchKeyLocked()` creates a new http.Client per call to override timeout — consider using `http.NewRequestWithContext()` with a deadline instead, avoiding client copy
- When `GetPublicKeyForKID` fetches but the AI server returns a different kid than requested, a debug log would help troubleshoot misconfigured AI servers

File: pkg/signing/keymanager.go [hunk 25]
```diff
@@ -0,0 +1,234 @@
...
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.publicKey = ed25519.PublicKey(keyBytes)
+	m.keyID = keyResp.KeyID
+	m.lastFetch = time.Now()
+
+	if keyResp.ExpiresAt != "" {
+		if t, err := time.Parse(time.RFC3339, keyResp.ExpiresAt); err == nil {
+			m.expiresAt = t
```
Comment (geunwoonoh) [inline pkg/signing/keymanager.go:222]: Before storing `keyResp.ExpiresAt`, it seems necessary to validate whether it is already expired. 
Although unlikely, if alpacon returns an already expired `expires_at`, the next `GetPublicKey()` call will trigger another fetch. This could potentially lead to an infinite fetch loop.

Therefore, if alpacon responds with an expired `expires_at`, it would be safer to either handle it with a TTL fallback or return an error. 
What do you think?

File: pkg/signing/verifier.go [hunk 26]
```diff
@@ -0,0 +1,123 @@
...
+		Line:      cmd.Line,
+		ServerID:  serverID,
+		Shell:     cmd.Shell,
+		Timestamp: cmd.AnalyzedAt,
+		Username:  cmd.User,
+	}
+	var buf bytes.Buffer
+	enc := json.NewEncoder(&buf)
+	enc.SetEscapeHTML(false)
+	_ = enc.Encode(p)
```
Comment (geunwoonoh) [inline pkg/signing/verifier.go:49]: Since all fields in `signingPayload` are `strings`, the likelihood of failure is very low. 
However, it would be good to return `nil` or a distinct error signature in case an error does occur.

File: pkg/signing/keymanager.go [hunk 27]
```diff
@@ -0,0 +1,234 @@
...
+	if err != nil {
+		return fmt.Errorf("failed to create request: %w", err)
+	}
+
+	client := &http.Client{
+		Transport:     m.client.Transport,
+		CheckRedirect: m.client.CheckRedirect,
+		Jar:           m.client.Jar,
+		Timeout:       m.fetchTimeout,
+	}
```
Comment (geunwoonoh) [inline pkg/signing/keymanager.go]: Creating a new HTTP client on every call to `fetchKeyLocked()` seems unnecessary.
Was this an intentional design choice?

File: pkg/signing/verifier.go [hunk 28]
```diff
@@ -0,0 +1,123 @@
...
+			i += 5 // skip past uXXXX
+		} else {
+			buf.WriteByte(data[i])
+		}
+	}
+	return buf.Bytes()
+}
+
+// VerifyCommand verifies the Ed25519 signature on a command.
+func VerifyCommand(cmd *protocol.Command, serverID string, publicKey ed25519.PublicKey) error {
```
Comment (geunwoonoh) [inline pkg/signing/verifier.go:96]: VerifyCommand accepts a bare ed25519.PublicKey, while Command.KeyID was added
to protocol.Command in this PR. The caller must use cmd.KeyID with
KeyManager.GetPublicKeyForKID to resolve the key before calling VerifyCommand,
but this contract is implicit.

Injecting KeyManager into VerifyCommand would couple verification logic with
key retrieval. The current design intentionally keeps VerifyCommand as a pure
function that is agnostic to key source (KID-based lookup, hardcoded key, test
fixture, etc.).

Recommend adding a doc comment to make the usage contract explicit:

    // VerifyCommand verifies the Ed25519 signature on a command.
    // The caller is responsible for resolving the public key, typically
    // by calling KeyManager.GetPublicKeyForKID(cmd.KeyID).

This turns the implicit contract into an explicit one without sacrificing
the function's flexibility

File: pkg/signing/keymanager_test.go [hunk 29]
```diff
@@ -0,0 +1,375 @@
...
+
+	if km.keyID != "key-test-123" {
+		t.Errorf("expected key ID 'key-test-123', got '%s'", km.keyID)
+	}
+}
+
+func TestKeyManager_CacheHit(t *testing.T) {
+	pub, _, _ := ed25519.GenerateKey(nil)
+
+	fetchCount := 0
```
Comment (geunwoonoh) [inline pkg/signing/keymanager_test.go]: fetchCount is incremented inside the httptest handler goroutine without
synchronization. While tests currently run with -p 1, this is a real data
race detectable by -race. If the sequential constraint is ever removed,
these tests break. Fix: use atomic.AddInt32 — one-line change per test.

File: pkg/signing/verifier.go [hunk 30]
```diff
@@ -0,0 +1,123 @@
...
+	var buf bytes.Buffer
+	buf.Grow(len(data))
+	for i := 0; i < len(data); i++ {
+		if data[i] == '\\' && i+5 < len(data) && data[i+1] == 'u' &&
+			(string(data[i+2:i+6]) == "2028" || string(data[i+2:i+6]) == "2029") {
+			// Count consecutive preceding backslashes
+			backslashCount := 0
+			for j := i - 1; j >= 0 && data[j] == '\\'; j-- {
+				backslashCount++
+			}
```
Comment (geunwoonoh) [inline pkg/signing/verifier.go:74]: The preceding backslash count is computed against the input buffer, not the
output buffer. Since `unescapeLineSeparators` copies bytes one at a time without
removing them from the input, the count is correct — but this correctness
depends on the input remaining unmodified during iteration. Add edge case
tests for `\\\\u2028` and `\\\\\\u2028` to lock this down.

**Comment by junho226** [comment]
@eunyoung14 Thanks for the review! Addressed both minor notes:

1. **`fetchKeyLocked()` client copy** — Replaced with `http.NewRequestWithContext()` using a deadline context. No more per-call client creation.
2. **Debug log for kid mismatch** — Added a debug log in `GetPublicKeyForKID` when the AI server returns a different kid than requested, logging both `requested_kid` and `cached_kid`.

File: pkg/signing/keymanager.go (hunk 25 above)
Comment (junho226) [reply pkg/signing/keymanager.go:222]: This scenario cannot cause an infinite loop. The call path is always linear: \`GetPublicKey()\` → \`Refresh()\` → \`fetchKeyLocked()\` → stores key → returns to caller. Even if the AI server returned an already-expired \`expires_at\`, each call would complete and return a key — it would just mean an HTTP fetch on every call rather than using the cache.

In practice, the AI server computes \`expires_at = now + KEY_TTL\` (default 90 days) at startup and refreshes it on key rotation, so it won't serve an already-expired value.

File: pkg/signing/verifier.go (hunk 26 above)
Comment (junho226) [reply pkg/signing/verifier.go:49]: All fields in \`signingPayload\` are \`string\` and the target is a \`bytes.Buffer\`, so \`Encode\` cannot fail — the only theoretical failure would be OOM, which causes a Go runtime panic regardless. Adding error handling here would create an untestable dead code path and a return signature that implies failure where none is possible.

File: pkg/signing/keymanager.go (hunk 27 above)
Comment (junho226) [reply pkg/signing/keymanager.go]: Good catch — it was done to apply a separate fetch timeout without mutating the shared client, but creating a new client per call is unnecessary. Fixed by using \`http.NewRequestWithContext()\` with a deadline context instead. The shared \`m.client\` is now used directly.

File: pkg/signing/verifier.go (hunk 28 above)
Comment (junho226) [reply pkg/signing/verifier.go:96]: Agreed — the pure function design is intentional, but the caller contract was implicit. Added the suggested doc comment to make it explicit.

File: pkg/signing/keymanager_test.go (hunk 29 above)
Comment (junho226) [reply pkg/signing/keymanager_test.go]: Fixed — replaced all \`fetchCount\` and \`callCount\` variables in test handlers with \`atomic.Int32\`. All tests now pass with \`-race\`.

File: pkg/signing/verifier.go (hunk 30 above)
Comment (junho226) [reply pkg/signing/verifier.go:74]: The correctness observation is right — the input buffer is immutable during iteration, so the preceding backslash count is always accurate.

The backslash counting logic is purely an odd/even parity check. The existing tests already cover both parities: 0 preceding backslashes (base case in \`TestBuildCanonicalPayload_LineSeparators\`) and 1 preceding backslash (\`TestBuildCanonicalPayload_LiteralBackslashU2028\`). Testing 2 or 4 preceding backslashes would exercise the same even-parity code path as 0, and 3 or 5 would exercise the same odd-parity path as 1 — no new branches are covered. Skipping the additional tests for this reason.

**Review by geunwoonoh — APPROVED** [review]
